### PR TITLE
Advanced Building Blocks -- Enumerables

### PR DESCRIPTION
**Pull request by blackpintz**
### **What the project does**
It creates enumerable methods that are identical to those of Ruby.

### ** Tasks completed**

- Created a module to house all the methods.
- Created 12 identical methods that are identical to the Ruby enumerable methods. Some include #each and #map.
-implemented #yield and call methods.

**How should this be manually tested?**
One option would be to clone this branch and the other would be to check out the [live webpage](https://repl.it/@blackpintz/EverlastingQuarrelsomeModes) for this project.

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    my_each do |x|
+      if block_given?
+        check = yield(x)
+        block_method(check, x, new_arr)
+      elsif num
+        param_method(x, num, new_arr)
+      end
+    end
+    new_arr.length && num.zero? && !block_given? ? length : new_arr.length
+  end
```
Comment (euqueme) [inline enumerables.rb:86]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_count

- [ ] my_count counts the number of items in enum that are equal to the item if an argument is given
```ruby
[5, 4, 1].my_count(0) == [5, 4, 1].count(0) => 0
```
- Yours is retrieving the wrong answer
image

- Real count
image

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      my_each do |x|
+        yield(x)
+        new_arr.push(yield(x))
+      end
+      new_arr
+    elsif prc
+      prc.call
+    else to_enum(:map)
+    end
+  end
```
Comment (euqueme) [inline enumerables.rb:100]: #### my_map 

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so let's work on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      while i < length
+        j = i - 1
+        x = self[i]
+        total = yield(x, total_arr[j])
+        total_arr.push(total)
+        i += 1
+      end
+    end
+    total_arr.pop
+  end
```
Comment (euqueme) [inline enumerables.rb:128]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is retrieving a NoMethodError
image

- Real inject
image

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
+module MyEnumerables
+  include Enumerable
+
+  def block_method(check, value, arr)
+    arr.push(value) if check
+  end
+
+  def param_method(value, param, arr)
+    arr.push(value) if param == value
+  end
```
Comment (euqueme) [inline enumerables.rb:10]: - [ ] both block_method and param_method aren't enumerable methods so they should be outside the MyEnumerables module

**Comment by blackpintz** [comment]
Fixed the requested changes.

**Pull request merged**